### ISSUE TITLE
[enhancement]: Installer requests python verions 3.11 or less - current version is 3.12.

### ISSUE THREAD

**Issue by ppboyle** [invoke-ai/InvokeAI#5674]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What should this feature add?

Why not let your installer respect the latest python version levels, vs having a user have to download older versions?

**Comment by hipsterusername** [comment]
This a dependency compatibility issue. PyTorch, a main machine learning library, currently only supports 3.11.

**Comment by Regnalf** [comment]
I have this issue too on a mac. But in my case it crashes the update.

![update_invoke_issue](https://github.com/invoke-ai/InvokeAI/assets/43412166/a904a88a-e00f-44af-9ed3-83dbbac08ffc)

Installed Version is 3.12.
Would pyenv help?

**Comment by UnconnectedBedna** [comment]
Use pyenv to install and the version of python you need, refer to your specific operating system on how to.

Make sure pyenv is enabled system wise
On arch based it can be something like:
```
pyenv install 3.10.6
pyenv local 3.10.6
pyenv shell 3.10.6
python -m venv venv
```

**Comment by tcmbackwards** [comment]
On windows you can use python launcher (py.exe) to choose the version which builds the venv during the installation. The venv will use the right version of python from then on.

Edit install.bat:

```
@rem -------------- Install and Configure ---------------

old:
call python .\lib\main.py 

new:
call py -3.11 .\lib\main.py 

pause
exit /b
```

**Comment by Coder-Harshit** [comment]
instead of virtualenv it would be better to revert to conda or miniconda to make it possible to have specific python install in env. i think so

**Comment by UnconnectedBedna** [comment]
> instead of virtualenv it would be better to revert to conda or miniconda

What do you think conda does?
It's just another way of creating a containerized environment. :)

Use what YOU feel comfortable with, the important part is you make it "containerized" so you can control what python version is used.

**Comment by leder11011** [comment]
> Use pyenv to install and the version of python you need, refer to your specific operating system on how to.
> 
> Make sure pyenv is enabled system wise On arch based it can be something like (from the directory where you run the application):
> 
> ```
> pyenv install 3.11.6
> pyenv local 3.11.6
> pyenv shell 3.11.6
> python -m venv venv
> ```

I followed the steps on manjaro - python 3.12 before, 3.11.6 after installation.
But I still get an error running the installer:

`/home/leder/invokeai/.venv/bin/python: error while loading shared libraries: libpython3.11.so.1.0: cannot open shared object file: No such file or directory`

**Comment by eboye** [comment]
> > Use pyenv to install and the version of python you need, refer to your specific operating system on how to.
> > Make sure pyenv is enabled system wise On arch based it can be something like (from the directory where you run the application):
> > ```
> > pyenv install 3.11.6
> > pyenv local 3.11.6
> > pyenv shell 3.11.6
> > python -m venv venv
> > ```
> 
> I followed the steps on manjaro - python 3.12 before, 3.11.6 after installation. But I still get an error running the installer:
> 
> `/home/leder/invokeai/.venv/bin/python: error while loading shared libraries: libpython3.11.so.1.0: cannot open shared object file: No such file or directory`

same here

**Comment by UnconnectedBedna** [comment]
No you did not follow the instructions quoted here, because if you did, the path would not be a hidden directory (.venv) but rather a directory called venv.

**Comment by JamesClarke7283** [comment]
It appears pytorch now supports 3.12, so you can remove the <3.11 requirement now.
https://pypi.org/project/torch/#files

**Comment by Chiaki** [comment]
**Please drop Python 3.11 and add support of 3.12 or 3.13 already!**
I know that security support of 3.11 will be till 2027, but `Python 3.11 isn't receiving regular bug fixes anymore, and binary installers are no longer provided for it. Python 3.11.9 was the last full bugfix release of Python 3.11 with binary installers.`
So on Windows platform it already outdated and there's no binaries to keep it secure. Since April 2 2024, 2 security patches were released.

**Comment by UnconnectedBedna** [comment]
Why though?
Just to make life as difficult as possible for the devs?
What advantages would be gained by changing?

As per install instructions, you are not required to install any kind of python on your system.
And even if you make a manual install, you are still recommended to use [uv](https://docs.astral.sh/uv/getting-started/installation/#standalone-installer)
https://invoke-ai.github.io/InvokeAI/installation/manual/#manual-install
> We'll use [uv](https://github.com/astral-sh/uv) to install python and create a virtual environment, then install the invokeai package. uv is a modern, very fast alternative to pip.

You don't need any installation of python on windows.

And if you use `pyenv`, I am not 100% sure, but I feel pretty confident that as long as you keep it updated, it will keep you safe. (correct me if I am wrong here)

**Comment by Chiaki** [comment]
First of all, this is from InvokeAI v5.4.2 installer `readme.txt`: `You will need to install Python 3.10 or higher for this installer to work.` Python 3.12 is **higher** than 3.10, right? But InvokeAI don't want to install with it. Of course I don't want to make devs life more difficult, but I can't upgrade InvokeAI after I updated my python to 3.12 and this is a problem.

I don't know why, but InvokeAI devs chosed to use 'system' python instead of 'embedded' as in ComfyUI. And in that case, I'm as user, expect to see a current release with active support, not legacy without binaries which means unpatched vulnerabilities in system. I know there's a tons of dependencies etcetera, but I believe that improvements and security should be in priority too.

As for `uv`, well, devs suggest to use the standalone installer method, but... `powershell -ExecutionPolicy ByPass -c "irm https://astral.sh/uv/install.ps1 | iex"` Executing unknown script, from unknown host, with bypassing execution policies?! No, tnx.

I think the best solution will be to have 'embedded' python, if it is impossible to maintain project with current active version of Python.

p.s.
No, I have python on my system for other AI stuff like text and video gen & etc. And none of them had issues with 3.12. except InvokeAI.

**Comment by UnconnectedBedna** [comment]
> I don't know why, but InvokeAI devs chosed to use 'system' python instead of 'embedded' as in ComfyUI

No they don't, literally in my last post:
> We'll use [uv](https://github.com/astral-sh/uv) to install python and create a virtual environment, then install the invokeai package. uv is a modern, very fast alternative to pip.

If you CHOOSE to use a native "system" installation of python instead of creating environments, it's ON YOU.
I don't think anybody recommends running local AI in that way.

If you decide to install stuff in YOUR way and say "no thanks" to the supported method, well, why are you typing here?
Your installation is unsupported anyway.

I'm now going to go use my python3.10.10 to generate FLUX images using forge... 😇

**Comment by Chiaki** [comment]
> If you CHOOSE to use a native "system" installation of python instead of creating environments, it's ON YOU.
No, I didn't CHOOSE it. Version 3 or earlier, I don't remember which it was, required 3.11 so I downloaded and installed it according to the manual. Because at that moment it was active support version. And after Python devs stopped providing binaries, I updated to 3.12 and couldn't update my InvokeAI.

If in 5.5.0 I can choose between 'embedded' and 'system' verions of python, that's great. But something tells me, that it still require 3.11 and will not update my version to latest. But I'll try.

**Comment by UnconnectedBedna** [comment]
> so I downloaded and installed it according to the manual.

Witch manual is that?
There is no mention of python anywhere
https://invoke-ai.github.io/InvokeAI/installation/quick_start

Remove all previous installations.
Reinstall using the correct method linked to above.
If THAT fails you should create a separate bug report here on the github.

**Comment by Chiaki** [comment]
> Witch manual is that?
> There is no mention of python anywhere

*sigh* Sir, why are you referring to latest version of manual after I said, that when I was installing InvokeAI, there was no such manual. Just check `Quick start` in README.md (commit 3cba535) and you will understand what I mean.

As follow up: yes, InvokeAI 5.5.0 finally uses 'embedded' (venv) python and update was successful. Yay.

**Comment by UnconnectedBedna** [comment]
You reference a readme file, comitted may 15 last year, that adds the word "updating"...
I know you grasp for anything to be correct.
The correct move is to admit you did not understand what you were doing.

> As follow up: yes, InvokeAI 5.5.0 finally uses 'embedded' (venv) python and update was successful. Yay.

v2.0 was released oct 10, 2022
https://github.com/invoke-ai/InvokeAI/blob/v2.0.0/docs/installation/INSTALL_WINDOWS.md

> Conda
> Install Anaconda3 (miniconda3 version) from [here](https://docs.anaconda.com/anaconda/install/windows/)

Very strange they instruct the usage of conda, an application that sets up venv:s...

Please just drop this.

**Comment by Chiaki** [comment]
> You reference a readme file, comitted may 15 last year, that adds the words "and Updates"... ok?
> I know you grasp for anything to try to be correct.

Yes. Because this was when I started using InvokeAI. And, thanks for the link, because I figured out how to navigate to what I was looking for. (: 

1. https://github.com/invoke-ai/InvokeAI/blob/v4.2.0/docs/installation/010_INSTALL_AUTOMATED.md - Getting installer, running installer, running app. Nothing about venv here.
2. Let's see [installation requirements](https://github.com/invoke-ai/InvokeAI/blob/v4.2.0/docs/installation/INSTALLATION.md#installation-requirements) -- First line: "Before installing, review the [installation requirements](https://github.com/invoke-ai/InvokeAI/blob/v4.2.0/docs/installation/INSTALL_REQUIREMENTS.md) to ensure your system is set up properly." ...follow the link.
3. [Python section](https://github.com/invoke-ai/InvokeAI/blob/v4.2.0/docs/installation/INSTALL_REQUIREMENTS.md#installing-python-windows) "Install python 3.11 with [an official installer](https://www.python.org/downloads/release/python-3118/)."

You know what it means, right? **Official InvokeAI v4.2.0 guide tells user to download and install Python as a 'system' instead of 'embedding' it.** And ofc, after I updated 3.11 to 3.12 I wasn't able to update to v5.4.2 (latest at that moment).

If Python devs continued provide binaries for 3.11 - there was no reason to update my 'system' Python to 3.12;
If InvokeAI devs used embedded Python since 4.2.x -- there was no problem with update, even with 'system' 3.12 or higher;
If InvokeAI devs keept support of all versions since 3.10 and higher (as they mentioned in readme) -- there was no problem with update too.

But in my case none of this happened. And I've got an error. I'm glad that in v5.5.0 they finally decided to embed Python and I was able to update, finally.

Please, don't think that if your experience or knowledge is better or different, all others just grasp for anything to try to be correct.

**Comment by UnconnectedBedna** [comment]
> Please, don't think that if your experience or knowledge is better or different

I don't think. I KNOW better. Lets analyze what you linked to...

> Before installing, review the [installation requirements](https://github.com/invoke-ai/InvokeAI/blob/v4.2.0/docs/installation/INSTALL_REQUIREMENTS.md) to ensure your system is set up properly.

Ok, lets read that:

> Invoke requires python 3.10 or 3.11. If you don't already have one of these versions installed, we suggest installing 3.11, as it will be supported for longer.

And yes, the automatic installation ustilizes a virtual python environment one way or another.
No further comment needed for this rambling.
